src/java.base/share/classes/com/sun/crypto/provider/TlsKeyMaterialGenerator.java

```diff
@@ -196,82 +201,97 @@ private SecretKey engineGenerateKey0() throws GeneralSecurityException {
-                                ivLength << 1, md5, sha);
-                    System.arraycopy(block, 0, tmp, 0, ivLength);
+                    byte[] tmp = new byte[ivLength];
+
+                    System.arraycopy(keyBlock, ofs, tmp, 0, ivLength);
+                    ofs += ivLength;
                     clientIv = new IvParameterSpec(tmp);
-                    System.arraycopy(block, ivLength, tmp, 0, ivLength);
+
+                    System.arraycopy(keyBlock, ofs, tmp, 0, ivLength);
+                    ofs += ivLength;
                     serverIv = new IvParameterSpec(tmp);
                 }
             } else {
-                // SSLv3
-                byte[] tmp = new byte[expandedKeyLength];
-
-                md5.update(clientKeyBytes);
-                md5.update(clientRandom);
-                md5.update(serverRandom);
-                System.arraycopy(md5.digest(), 0, tmp, 0, expandedKeyLength);
-                clientCipherKey = new SecretKeySpec(tmp, alg);
-
-                md5.update(serverKeyBytes);
-                md5.update(serverRandom);
-                md5.update(clientRandom);
-                System.arraycopy(md5.digest(), 0, tmp, 0, expandedKeyLength);
-                serverCipherKey = new SecretKeySpec(tmp, alg);
+                // if exportable suites, calculate the alternate
+                // cipher key expansion and IV generation
+                if (protocolVersion >= 0x0302) {
+                    // TLS 1.1+
+                    throw new RuntimeException(
+                            "Internal Error:  TLS 1.1+ should not be negotiating" +
+                                    "exportable ciphersuites");
+                } else if (protocolVersion == 0x0301) {
+                    // TLS 1.0
+                    byte[] seed = concat(clientRandom, serverRandom);
+
+                    byte[] tmp = doTLS10PRF(clientKeyBytes,
+                            LABEL_CLIENT_WRITE_KEY, seed, expandedKeyLength, md5, sha);
+                    clientCipherKey = new SecretKeySpec(tmp, alg);
+                    Arrays.fill(tmp, (byte) 0);
+
+                    tmp = doTLS10PRF(serverKeyBytes, LABEL_SERVER_WRITE_KEY, seed,
+                            expandedKeyLength, md5, sha);
+                    serverCipherKey = new SecretKeySpec(tmp, alg);
+                    Arrays.fill(tmp, (byte) 0);
 
-                if (ivLength != 0) {
-                    tmp = new byte[ivLength];
+                    if (ivLength != 0) {
+                        tmp = new byte[ivLength];
+                        byte[] block = doTLS10PRF(null, LABEL_IV_BLOCK, seed,
+                                ivLength << 1, md5, sha);
+                        System.arraycopy(block, 0, tmp, 0, ivLength);
+                        clientIv = new IvParameterSpec(tmp);
+                        System.arraycopy(block, ivLength, tmp, 0, ivLength);
+                        serverIv = new IvParameterSpec(tmp);
```

Comment:
Again, consider using IvParameterSpec(byte[], int, int) and get rid of tmp?

Reply:
Yes.